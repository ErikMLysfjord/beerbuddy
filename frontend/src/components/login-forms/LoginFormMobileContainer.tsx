import { Form, Input, Button } from "antd";
import { ValidateErrorEntity } from "rc-field-form/lib/interface";
import styles from "./LoginFormMobile.module.css";

//? What is this
type FieldType = {
  username?: string;
  password?: string;
};

interface LoginFormMobileProps {
  onFinishFailed: (errorInfo: ValidateErrorEntity) => void;
  saveUser: (string: { username: string }) => void;
}

/**
 * Mobile version of the login form component.
 * Contains the actual form that is rendered in LoginFormMobile.tsx.
 * @param onFinishFailed - function that is called when the form is submitted and fails validation
 * @param saveUser - function that is called when the form is submitted and passes validation
 * @returns
 */
const LoginFormMobile = ({
  onFinishFailed,
  saveUser,
}: LoginFormMobileProps) => {
  return (
    <section>
      <h1 className={styles.loginFormHeaderMobile}>Log in</h1>
      <Form
        name="basic"
        className={styles.loginFormMobile}
        layout="vertical"
        initialValues={{ remember: true }}
        onFinish={(values) => {
          saveUser({ username: values.username });
        }}
        onFinishFailed={onFinishFailed}
        autoComplete="off"
      >
        <Form.Item<FieldType>
          label="Username"
          name="username"
          className={styles.loginFormUsername}
          rules={[{ required: true, message: "Please input your username!" }]}
        >
          <Input />
        </Form.Item>

        <Form.Item>
          <Button
            type="primary"
            htmlType="submit"
            className={styles.loginButton}
          >
            Submit
          </Button>
        </Form.Item>
      </Form>
    </section>
  );
};

export default LoginFormMobile;
