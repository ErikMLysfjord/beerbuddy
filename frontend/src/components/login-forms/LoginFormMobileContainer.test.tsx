import { describe, it, vi, expect } from "vitest";
import {
  render,
  waitFor,
  screen,
  fireEvent,
  act,
} from "@testing-library/react";
import LoginFormMobileContainer from "./LoginFormMobileContainer";
import { axe } from "jest-axe";

const onFinishFailed = vi.fn();
const saveUser = vi.fn();

describe("LoginFormMobileContainer", () => {
  it("is accessible", async () => {
    const { container } = render(
      <LoginFormMobileContainer
        onFinishFailed={onFinishFailed}
        saveUser={saveUser}
      />
    );

    expect(await axe(container)).toHaveNoViolations();
  });

  it("matches snapshot", async () => {
    const { container } = render(
      <LoginFormMobileContainer
        onFinishFailed={onFinishFailed}
        saveUser={saveUser}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it("renders without crashing", () => {
    const { container } = render(
      <LoginFormMobileContainer
        onFinishFailed={onFinishFailed}
        saveUser={saveUser}
      />
    );
    expect(container).toBeTruthy();
  });

  it("handles submit", async () => {
    render(
      <LoginFormMobileContainer
        onFinishFailed={onFinishFailed}
        saveUser={(string: { username: string }) => saveUser(string.username)}
      />
    );

    act(() => {
      fireEvent.change(screen.getByRole("textbox"), {
        target: { value: "test" },
      });
      screen.getByRole("button").click();
    });

    await waitFor(() => {
      expect(saveUser).toHaveBeenCalledTimes(1);
      expect(saveUser).toHaveBeenCalledWith("test");
    });
  });
});
