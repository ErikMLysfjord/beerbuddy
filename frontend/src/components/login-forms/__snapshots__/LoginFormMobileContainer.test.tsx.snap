// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginFormMobileContainer > matches snapshot 1`] = `
<div>
  <section
    aria-label="Log in form"
  >
    <h1
      class="_loginFormHeaderMobile_b69d51"
    >
      Log in
    </h1>
    <form
      autocomplete="off"
      class="ant-form ant-form-vertical css-dev-only-do-not-override-pr0fja _loginFormMobile_b69d51"
      id="basic"
    >
      <div
        class="ant-form-item _loginFormUsername_b69d51 css-dev-only-do-not-override-pr0fja"
      >
        <div
          class="ant-row ant-form-item-row css-dev-only-do-not-override-pr0fja"
        >
          <div
            class="ant-col ant-form-item-label css-dev-only-do-not-override-pr0fja"
          >
            <label
              class="ant-form-item-required"
              for="basic_username"
              title="Username"
            >
              Username
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control css-dev-only-do-not-override-pr0fja"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <input
                  aria-required="true"
                  class="ant-input css-dev-only-do-not-override-pr0fja"
                  id="basic_username"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-form-item css-dev-only-do-not-override-pr0fja"
      >
        <div
          class="ant-row ant-form-item-row css-dev-only-do-not-override-pr0fja"
        >
          <div
            class="ant-col ant-form-item-control css-dev-only-do-not-override-pr0fja"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <button
                  class="ant-btn css-dev-only-do-not-override-pr0fja ant-btn-primary _loginButton_b69d51"
                  type="submit"
                >
                  <span>
                    Submit
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </section>
</div>
`;
